static/js/components/SaveCandidateButton.jsx

```diff
@@ -0,0 +1,210 @@
+import React, { useState, useEffect, useRef } from "react";
+import PropTypes from "prop-types";
+import { useDispatch } from "react-redux";
+import Dialog from "@material-ui/core/Dialog";
+import DialogContent from "@material-ui/core/DialogContent";
+import DialogTitle from "@material-ui/core/DialogTitle";
+import Checkbox from "@material-ui/core/Checkbox";
+import Button from "@material-ui/core/Button";
+import ButtonGroup from '@material-ui/core/ButtonGroup';
+import FormControlLabel from "@material-ui/core/FormControlLabel";
+import ArrowDropDownIcon from "@material-ui/icons/ArrowDropDown";
+import ClickAwayListener from "@material-ui/core/ClickAwayListener";
+import Grow from "@material-ui/core/Grow";
+import Paper from "@material-ui/core/Paper";
+import Popper from "@material-ui/core/Popper";
+import MenuItem from "@material-ui/core/MenuItem";
+import MenuList from "@material-ui/core/MenuList";
+import { useForm, Controller } from "react-hook-form";
+
+import * as sourceActions from "../ducks/source";
+import FormValidationError from "./FormValidationError";
+
+
+const options = ["Save as Source", "Save (select groups)"];
+
+const SaveCandidateButton = ({ candidate, userGroups }) => {
+  // Dialog logic:
+
+  const dispatch = useDispatch();
+  const [dialogOpen, setDialogOpen] = useState(false);
+
+  const { handleSubmit, errors, reset, control, getValues } = useForm();
+
+  useEffect(() => {
+    reset({
+      group_ids: userGroups.map((userGroup) => candidate.passing_group_ids.includes(userGroup.id))
+    });
+  }, [reset, userGroups, candidate]);
+
+  const handleClickOpenDialog = () => {
```

Comment:
We're starting to see numerous dialog boxes, so would it make sense to farm a dialog box out into a separate component that only knows how to open and close itself?

Reply:
@stefanv actually, the element whose `onClick` triggers the opening of the dialog needs to be defined within the dialog (because it needs to call the dialog's `handleOpen`), which is why I haven't factored this out already. Can you think of a (simple) way to do that?

Reply:
After reading about this exact use case in the offical React docs (https://reactjs.org/docs/refs-and-the-dom.html#when-to-use-refs), they recommend handling the open/close logic outside of the dialog (if the dialog is a child component) and passing `isOpen` as a prop to the dialog component. So we could factor out the `<Dialog><DialogContent>` part, but that doesn't spare much code, as the logic needs to be implemented in each instance.

Reply:
@stefanv can we move this discussion to an issue and resolve for now? This doesn't seem like an obvious should-do considering the above, and I don't want to drag out the review process unnecessarily.

Reply:
Certainly; thanks for looking into this!